crates/bevy_transform/src/systems.rs

```diff
@@ -2,74 +2,93 @@ use crate::components::{GlobalTransform, Transform};
+        // large trees are not clumped together.
+        1,
+        |(entity, children, transform, mut changed, children_changed, mut global_transform)| {
+            if changed {
+                *global_transform = GlobalTransform::from(*transform);
+            }
 
-        if let Some((children, changed_children)) = children {
             // If our `Children` has changed, we need to recalculate everything below us
-            changed |= changed_children;
-            for child in children {
+            changed |= children_changed;
+
+            for child in children.iter() {
                 let _ = propagate_recursive(
                     &global_transform,
-                    &mut transform_query,
+                    &transform_query,
+                    &parent_query,
                     &children_query,
-                    *child,
                     entity,
+                    *child,
                     changed,
                 );
             }
-        }
-    }
+        },
+    );
 }
 
 fn propagate_recursive(
     parent: &GlobalTransform,
-    transform_query: &mut Query<(
-        &Transform,
-        Changed<Transform>,
-        &mut GlobalTransform,
-        &Parent,
-    )>,
+    unsafe_transform_query: &Query<
+        (&Transform, Changed<Transform>, &mut GlobalTransform),
+        With<Parent>,
+    >,
+    parent_query: &Query<&Parent>,
     children_query: &Query<(&Children, Changed<Children>), (With<Parent>, With<GlobalTransform>)>,
-    entity: Entity,
     expected_parent: Entity,
+    entity: Entity,
     mut changed: bool,
     // We use a result here to use the `?` operator. Ideally we'd use a try block instead
 ) -> Result<(), ()> {
-    let global_matrix = {
-        let (transform, transform_changed, mut global_transform, child_parent) =
-            transform_query.get_mut(entity).map_err(drop)?;
-        // Note that for parallelising, this check cannot occur here, since there is an `&mut GlobalTransform` (in global_transform)
+    if let Ok(actual_parent) = parent_query.get(entity) {
         assert_eq!(
-            child_parent.get(), expected_parent,
+            actual_parent.get(), expected_parent,
             "Malformed hierarchy. This probably means that your hierarchy has been improperly maintained, or contains a cycle"
         );
+    } else {
+        panic!("Propagated child for {:?} has no Parent component!", entity);
+    }
+
+    // SAFE: With the check that each child has one and only one parent, each child must be globally unique within the
+    // hierarchy. Because of this, it is impossible for this query to have aliased mutable access to the same GlobalTransform.
+    // Any malformed hierarchy will cause a panic due to the above check.
+    let global_matrix = unsafe {
+        let (transform, transform_changed, mut global_transform) =
+            unsafe_transform_query.get_unchecked(entity).map_err(drop)?;
```

Comment:
Letting the function return `()` and doing `let ... else { return; };` would be a lot nicer than the interesting use of `.map_err(drop)?`.
(I know it's not your handiwork)
The unsafe block is also a bit big.

Reply:
Done.